packages/devtools_app/lib/src/inspector/flutter/inspector_tree_flutter.dart

```diff
@@ -0,0 +1,505 @@
+        );
+      },
+      child: InspectorRowContent(
+        row: widget.row,
+        expandAnimation: expandAnimation,
+        controller: widget.inspectorTreeState.controller,
+        onToggle: () {
+          setExpanded(!isExpanded);
+        },
+      ),
+    );
+  }
+
+  @override
+  bool get isExpanded => widget.node.isExpanded;
+
+  @override
+  void onExpandChanged(bool expanded) {
+    setState(() {
+      final row = widget.row;
+      if (expanded) {
+        widget.inspectorTreeState.controller.onExpandRow(row);
+      } else {
+        widget.inspectorTreeState.controller.onCollapseRow(row);
+      }
+    });
+  }
+
+  @override
+  bool shouldShow() => widget.node.shouldShow;
+}
+
+class InspectorTreeControllerFlutter extends Object
+    with InspectorTreeController, InspectorTreeFixedRowHeightController {
+  /// Client the controller notifies to trigger changes to the UI.
+  InspectorControllerClient get client => _client;
+  InspectorControllerClient _client;
+  set client(InspectorControllerClient value) {
+    if (_client == value) return;
+    // Do not set a new client if there is still an old client.
+    assert(value == null || _client == null);
+    _client = value;
+
+    if (config.onClientActiveChange != null) {
+      config.onClientActiveChange(value != null);
+    }
+  }
+
+  @override
+  InspectorTreeNode createNode() => InspectorTreeNode();
+
+  @override
+  Rect getBoundingBox(InspectorTreeRow row) {
+    // For future reference: the bounding box likely needs to be in terms of
+    // positions after the current animations are complete so that computations
+    // to start animations to show specific widget scroll to where the target
+    // nodes will be displayed rather than where they are currently displayed.
+    return Rect.fromLTWH(
+      getDepthIndent(row.depth),
+      getRowY(row.index),
+      rowWidth,
+      rowHeight,
+    );
+  }
+
+  @override
+  void scrollToRect(Rect targetRect) {
+    client?.scrollToRect(targetRect);
+  }
+
+  @override
+  void setState(VoidCallback fn) {
+    fn();
+    client?.onChanged();
+  }
+
+  /// Width each row in the tree should have ignoring its indent.
+  ///
+  /// Content in rows should wrap if it exceeds this width.
+  final double rowWidth = 1200;
+
+  /// Maximum indent of the tree in pixels.
+  double _maxIndent;
+
+  double get maxRowIndent {
+    if (lastContentWidth == null) {
+      double maxIndent = 0;
+      for (int i = 0; i < numRows; i++) {
+        final row = getCachedRow(i);
+        if (row != null) {
+          maxIndent = max(maxIndent, getDepthIndent(row.depth));
+        }
+      }
+      lastContentWidth = maxIndent + maxIndent;
+      _maxIndent = maxIndent;
+    }
+    return _maxIndent;
+  }
+}
+
+abstract class InspectorControllerClient {
+  void onChanged();
+
+  void scrollToRect(Rect rect);
+}
+
+class _InspectorTreeState extends State<InspectorTree>
+    implements InspectorControllerClient {
```

Comment:
This is an interesting take on the pattern of defining listeners.  It's different from the standard approach of handing tearoffs to the controller call back when asking for a change, and it gave me pause at first. 

Semantically the approaches are the same.  Go for it.

Reply:
Can you give me a concrete example of the pattern you are suggesting? Happy to also just see a CL refactoring this to the traditional pattern. I didn't mean to do something non-standard here.

Reply:
```dart
controller = Controller(
  onSomething: _doSomething,
  onSomethingElse: _doSomethingElse,
);
```

or, for a longer-lived controller:

```dart
controller.addListener(
  onSomething: _doSomething,
  onSomethingElse: _doSomethingElse,
);
```

But your approach here with a defined interface is equivalent.

So long as we are initState()ing and dispose()ing properly, there's nothing to be concerned about.